File: Lab_2/Lab_2/Lab_2.cpp
#include <iostream>
#include <string>
#include <vector>

using namespace std;

class Client {
private:
    int client_id;
    string name;
    string surname;
    string phone_number;

public:
    Client(int id, string nm, string srn, string phone)
        : client_id(id), name(nm), surname(srn), phone_number(phone) {}

    string getName() const { return name; }
    void setName(const string& nm) { name = nm; }

    string getPhoneNumber() const { return phone_number; }
    void setPhoneNumber(const string& phone) { phone_number = phone; }

    void displayClientInfo() const {
        cout << "Client: " << name << " " << surname << ", Phone: " << phone_number << endl;
    }

    void updatePhoneNumber(const string& new_phone) {
        phone_number = new_phone;
        cout << "Phone number updated to: " << phone_number << endl;
    }
};

class Payment {
public:
    virtual void processPayment(double amount) const {
        cout << "Processing payment of $" << amount << endl;
    }
};

class Device {
private:
    int device_id;
    string type;
    string brand;

public:
    Device(int id, string tp, string brd)
        : device_id(id), type(tp), brand(brd) {}

    void displayDeviceInfo() const {
        cout << "Device: " << brand << " " << type << endl;
    }
};

class RepairApplication : public Client, public Device, public Payment {
private:
    int application_id;
    string status;
    double planned_price;

public:
    RepairApplication(int app_id, string stat, double price, int client_id, string name, string surname, string phone,
        int device_id, string type, string brand)
        : Client(client_id, name, surname, phone), Device(device_id, type, brand),
        application_id(app_id), status(stat), planned_price(price) {}

    void displayApplicationInfo() const {
        displayClientInfo();
        displayDeviceInfo();
        cout << "Application ID: " << application_id << ", Status: " << status
            << ", Planned Price: $" << planned_price << endl;
    }

    void processPayment(double amount) const override {
        cout << "Payment for repair application " << application_id << ": $" << amount << endl;
    }
};

class Worker : public Client {
private:
    int worker_id;
    string position;

public:
    Worker(int id, string nm, string srn, string phone, int w_id, string pos)
        : Client(id, nm, srn, phone), worker_id(w_id), position(pos) {}

    void displayWorkerInfo() const {
        displayClientInfo();
        cout << "Worker ID: " << worker_id << ", Position: " << position << endl;
    }

    void assignRepair(int application_id) {
        cout << "Worker assigned to repair application " << application_id << endl;
    }
};

int main() {
    Client client1(1, "John", "Doe", "123-456-789");
    client1.displayClientInfo();
    client1.updatePhoneNumber("987-654-321");

    Worker worker1(2, "Michael", "Smith", "987-654-123", 101, "Technician");
    worker1.displayWorkerInfo();

    RepairApplication app1(1001, "In Progress", 250.75, 1, "John", "Doe", "987-654-321", 2001, "Laptop", "Dell");
    app1.displayApplicationInfo();

    RepairApplication app2(1002, "Completed", 150.50, 3, "Alice", "Johnson", "555-123-456", 2002, "Phone", "Samsung");
    app2.displayApplicationInfo();

    app1.processPayment(250.75);
    app2.processPayment(150.50);

    worker1.assignRepair(1001);

    return 0;
}

File: Lab_2/Lab_2/Lab_2.py
class Client:
    client_counter = 0  

    def __init__(self, client_id, name, surname, phone_number):
        self.client_id = client_id
        self.name = name
        self.surname = surname
        self.__phone_number = phone_number  
        Client.client_counter += 1  

    @staticmethod
    def client_count():
        return Client.client_counter  

    @property
    def phone_number(self):
        return self.__phone_number

    @phone_number.setter
    def phone_number(self, value):
        self.__phone_number = value

    def display_client_info(self):
        print(f"Client: {self.name} {self.surname}, Phone: {self.__phone_number}")

    def update_phone_number(self, new_phone):
        self.phone_number = new_phone
        print(f"Phone number updated to: {self.__phone_number}")


class Payment:
    def process_payment(self, amount):
        print(f"Processing payment of ${amount}")


class Device:
    def __init__(self, device_id, type, brand):
        self.device_id = device_id
        self.type = type
        self.brand = brand

    def display_device_info(self):
        print(f"Device: {self.brand} {self.type}")


class RepairApplication(Client, Device, Payment):
    def __init__(self, application_id, status, planned_price, client_id, name, surname, phone, device_id, type, brand):
        Client.__init__(self, client_id, name, surname, phone)
        Device.__init__(self, device_id, type, brand)
        self.application_id = application_id
        self.status = status
        self.planned_price = planned_price

    def display_application_info(self):
        self.display_client_info()
        self.display_device_info()
        print(f"Application ID: {self.application_id}, Status: {self.status}, Planned Price: ${self.planned_price}")

    def process_payment(self, amount):
        super().process_payment(amount)  
        print(f"Payment for repair application {self.application_id}: ${amount}")


class Worker(Client):
    def __init__(self, client_id, name, surname, phone, worker_id, position):
        super().__init__(client_id, name, surname, phone)
        self.worker_id = worker_id
        self.position = position

    def display_worker_info(self):
        self.display_client_info()
        print(f"Worker ID: {self.worker_id}, Position: {self.position}")

    def assign_repair(self, application_id):
        print(f"Worker assigned to repair application {application_id}")


class Technician(Payment):
    def __init__(self, tech_id, name, surname):
        self.tech_id = tech_id
        self.name = name
        self.surname = surname

    def display_technician_info(self):
        print(f"Technician: {self.name} {self.surname}, ID: {self.tech_id}")

    def process_payment(self, amount):
        super().process_payment(amount)
        print(f"Technician payment processed for ${amount}")


class RepairWorker(Worker, Technician):
    def __init__(self, client_id, name, surname, phone, worker_id, position, tech_id):
        Worker.__init__(self, client_id, name, surname, phone, worker_id, position)
        Technician.__init__(self, tech_id, name, surname)

    def display_repair_worker_info(self):
        self.display_worker_info()
        self.display_technician_info()


client1 = Client(1, "John", "Doe", "123-456-789")
client1.display_client_info()
client1.update_phone_number("987-654-321")

worker1 = Worker(2, "Michael", "Smith", "987-654-123", 101, "Technician")
worker1.display_worker_info()

app1 = RepairApplication(1001, "In Progress", 250.75, 1, "John", "Doe", "987-654-321", 2001, "Laptop", "Dell")
app1.display_application_info()

app2 = RepairApplication(1002, "Completed", 150.50, 3, "Alice", "Johnson", "555-123-456", 2002, "Phone", "Samsung")
app2.display_application_info()

app1.process_payment(250.75)
app2.process_payment(150.50)

worker1.assign_repair(1001)

tech_worker = RepairWorker(3, "Sarah", "Connor", "555-987-654", 102, "Senior Technician", 201)
tech_worker.display_repair_worker_info()
tech_worker.process_payment(300.00)

print(f"Total clients: {Client.client_count()}")  
